From 78b528c663d3dea5c0c51a6a1731db0e18bb451f Mon Sep 17 00:00:00 2001
From: Daniel Schwierzeck <daniel.schwierzeck@sphairon.com>
Date: Mon, 2 May 2011 13:23:19 +0200
Subject: [PATCH 1/5] schwierd-disable_locales

---
 src/qString.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/qString.c b/src/qString.c
index 54b3f24..3c9b53a 100644
--- a/src/qString.c
+++ b/src/qString.c
@@ -698,7 +698,7 @@ bool qStrIsIpv4Addr(const char *str) {
 	return true;
 }
 
-#ifdef __linux__
+#if 0
 #include <iconv.h>
 #endif
 
@@ -717,7 +717,7 @@ bool qStrIsIpv4Addr(const char *str) {
  * @endcode
  */
 char *qStrConvEncoding(const char *str, const char *fromcode, const char *tocode, float mag) {
-#ifdef __linux__
+#if 0
 	if(str == NULL) return NULL;
 
 	char *fromstr = (char*)str;
-- 
1.7.4.4

