Index: bridge-utils-1.2/libbridge/libbridge.h
===================================================================
--- bridge-utils-1.2.orig/libbridge/libbridge.h	2015-07-31 12:34:41.788264459 +0200
+++ bridge-utils-1.2/libbridge/libbridge.h	2015-07-31 12:36:18.046302793 +0200
@@ -20,7 +20,13 @@
 #define _LIBBRIDGE_H
 
 #include <net/if.h>
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,8,00)
 #include <linux/if_bridge.h>
+#else
+#include <netinet/ip6.h>
+#endif
 
 struct bridge_id
 {
