---
 src/ntfs-3g.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/src/ntfs-3g.c
+++ b/src/ntfs-3g.c
@@ -3737,14 +3737,14 @@ static fuse_fstype load_fuse_module(void
 	int i;
 	struct stat st;
 	pid_t pid;
-	const char *cmd = "/sbin/modprobe";
+	const char *cmd = "/sbin/insmod";
 	struct timespec req = { 0, 100000000 };   /* 100 msec */
 	fuse_fstype fstype;
 	
 	if (!stat(cmd, &st) && !geteuid()) {
 		pid = fork();
 		if (!pid) {
-			execl(cmd, cmd, "fuse", NULL);
+			execl(cmd, cmd, "/lib/modules/fuse.ko", NULL);
 			_exit(1);
 		} else if (pid != -1)
 			waitpid(pid, NULL, 0);
