From: Alexander Sack <asac@jwsdot.com>
Date: Thu, 18 Dec 2008 05:30:18 +0100
Subject: ppp-2.4.4rel/debian/patches/fix_linkpidfile

---
 pppd/main.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/pppd/main.c b/pppd/main.c
index 0c51e91..7d9d355 100644
--- a/pppd/main.c
+++ b/pppd/main.c
@@ -773,8 +773,7 @@ detach()
 	/* update pid files if they have been written already */
 	if (pidfilename[0])
 	    create_pidfile(pid);
-	if (linkpidfile[0])
-	    create_linkpidfile(pid);
+	create_linkpidfile(pid);
 	exit(0);		/* parent dies */
     }
     setsid();
--

